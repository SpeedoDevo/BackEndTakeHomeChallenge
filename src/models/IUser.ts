import IUserModel from './IUserModel'

export default interface IUser {
  getId(): number
  getName(): string
  getUsername(): string
  getAvatar(): string | undefined
  updateName(name: string): this
  updateAvatar(avatar?: string): this
  toObject(): IUserModel
}
