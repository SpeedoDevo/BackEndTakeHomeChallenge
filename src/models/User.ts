import IUser from './IUser'
import IListing from './IListing'

export default class User implements IUser {
  private id: number
  private name: string
  private listings: Array<IListing>

  constructor({ id, name, listings = [] }) {
    this.id = id
    this.name = name
    this.listings = listings
  }

  /**
   * Returns the UUID 
   */
  public getId() {
    return this.id
  }

  /**
   * Returns the name of this user
   */
  public getName() {
    return this.name
  }

  /**
   * Sets the name of this user
   */
  public updateName(name: string) {
    this.name = name
    return this
  }

  /**
   * Get listings
   */
  public getListings() {
    return this.listings
  }

  /**
   * Adds a new listing to this user
   */
  public addListing(listing: IListing) {
    this.listings.push(listing)
    return this
  }

  /**
   * Returns the Object version of this instance
   */
  public toObject() {
    return {
      id: this.getId(),
      name: this.getName(),
      listings: this.getListings()
    }
  }
}