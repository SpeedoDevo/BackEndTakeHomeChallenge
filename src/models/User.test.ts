import IUser from './IUser'
import User from './User'

describe('User', () => {
  let user: IUser

  beforeEach(() => {
    user = new User({
      id: 12,
      name: 'Joe',
      username: 'joe',
    })
  })

  describe('getId', () => {
    it('should get the right thing', () => {
      expect(user.getId()).toBe(12)
    })
  })
  describe('getName', () => {
    it('should get the right thing', () => {
      expect(user.getName()).toBe('Joe')
    })
  })
  describe('getUsername', () => {
    it('should get the right thing', () => {
      expect(user.getUsername()).toBe('joe')
    })
  })
  describe('getAvatar', () => {
    it('should get the right thing', () => {
      expect(user.getAvatar()).toBe(undefined)
    })
  })

  describe('updateName', () => {
    beforeEach(() => {
      user.updateName('Joey')
    })

    it('should update', () => {
      expect(user.getName()).toBe('Joey')
    })
  })

  describe('updateAvatar', () => {
    beforeEach(() => {
      user.updateAvatar('giphy.com/something-something')
    })

    it('should update', () => {
      expect(user.getAvatar()).toBe('giphy.com/something-something')
    })
  })

  describe('toObject', () => {
    it('should match snapshot', () => {
      expect(user.toObject()).toMatchSnapshot()
    })
  })
})
