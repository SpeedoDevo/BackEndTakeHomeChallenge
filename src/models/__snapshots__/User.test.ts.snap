// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User toObject should match snapshot 1`] = `
Object {
  "avatar": undefined,
  "id": 12,
  "name": "Joe",
  "username": "joe",
}
`;
